xdsl/transforms/eqsat_add_costs.py

```diff
@@ -0,0 +1,66 @@
+from xdsl.context import MLContext
+from xdsl.dialects import builtin, eqsat
+from xdsl.dialects.builtin import IntAttr
+from xdsl.ir import OpResult
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+
+class AddCostEclass(RewritePattern):
+    """
+    Add cost to each operator in an e-class.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: eqsat.EClassOp, rewriter: PatternRewriter):
+        for operand in op.operands:
+            if not isinstance(operand, OpResult):
+                # only add costs to operators
+                continue
+
+            # Add cost attribute to operators
+            operation = operand.op
+            operation.attributes["cost"] = IntAttr(1)
+
+
+class EqsatAddCosts(ModulePass):
+    """
+    Replace all eqsat.eclass operations in an MLIR program.
+
+    Input example:
+       ```mlir
+       func.func @test(%a : index, %b : index) -> (index) {
+            %a_eq = eqsat.eclass %a : index
+            %b_eq = eqsat.eclass %b : index
+            %c_ab = arith.addi %a_eq, %b_eq   : index
+            %c_ba = arith.addi %b_eq, %a_eq   : index
+            %c_eq = eqsat.eclass %c_ab, %c_ba : index
+            func.return %c_eq : index
+        }
+       ```
+    Output example:
+        ```mlir
+        func.func @test(%a : index, %b : index) -> (index) {
+           %c_ab = arith.addi %a, %b : index
+           func.return %c_ab : index
+        }
+        ```
+    """
+
+    name = "eqsat-add-costs"
+
+    def apply(self, ctx: MLContext, op: builtin.ModuleOp) -> None:
+        PatternRewriteWalker(
+            GreedyRewritePatternApplier(
+                [
+                    AddCostEclass(),
+                ]
+            ),  # list of rewrite patterns
```

Comment:
I'm not sure the comments are very informatice, I would recommend deleting them. Also, I would expect this pass to work the same without the greedy rewrite applier and with `apply_recursively` set to `False`

Reply:
Don't think my pass is actually working it only assigns costs to some operators - e.g. in the example the multiplier does not get assigned a cost? Any ideas why it's not matching on all the e-class operators?

Reply:
I think your pass works but the printer for some of the arith operations is faulty and doesn't print attributes

Reply:
See here:
https://github.com/xdslproject/xdsl/blob/7c86cfba1c8617f9da62b5a752222310008ec86d/xdsl/dialects/arith.py#L204-L210

Reply:
Nice thanks for the clarification!